solutions/usaco-813.mdx

```diff
@@ -5,54 +5,248 @@ title: Snow Boots
-			needed_step = max(needed_step,*next(it)-*prev(it));
-			indices.erase(it);
-			++tile_index;
+public class SnowBoots {
+	//BeginCodeSnip{Boot Class}
+	static class Boot {
+		public int maxDepth;
+		public int maxSteps;
+		public int index;
+		public Boot(int maxDepth, int maxSteps, int index) {
+			this.maxDepth = maxDepth;
+			this.maxSteps = maxSteps;
+			this.index = index;
 		}
-		ans[a[2]] = a[1] >= needed_step;
 	}
+	//EndCodeSnip
+
+	public static void main(String[] args) throws IOException {
+		Kattio io = new Kattio("snowboots");
+		
+		int tileNum = io.nextInt();
+		int bootNum = io.nextInt();
+
+		int[] tiles = new int[tileNum];
+		for (int t = 0; t < tileNum; t++) {
+			tiles[t] = io.nextInt();
+		}
+
+		Boot[] boots = new Boot[bootNum];
+		for (int i = 0; i < bootNum; i++) {
+			boots[i] = new Boot(io.nextInt(), io.nextInt(), i);
+		}
+		// sort the boots from greatest depth to least depth
+		Arrays.sort(boots, Comparator.comparingInt(b -> -b.maxDepth));
+
+		// list of tile indices, sorted by snow depth
+		Integer[] rawTilesByDepth = new Integer[tileNum - 2];
+		for (int i = 1; i < tileNum - 1; i++) {
+			rawTilesByDepth[i - 1] = i;
+		}
+		// we need the Integer class to use custom comparators
+		Arrays.sort(rawTilesByDepth, Comparator.comparingInt(t -> -tiles[t]));
+
+		// cast the Integers back to ints for performance reasons
+		int[] tilesByDepth = new int[tileNum - 2];
+		for (int i = 0; i < tilesByDepth.length; i++) {
+			tilesByDepth[i] = rawTilesByDepth[i];
+		}
+
+		TreeSet<Integer> validTiles = new TreeSet<>();
+		for (int t = 0; t < tileNum; t++) {
+			validTiles.add(t);
+		}
+
+		// the deepest tile that can still be traversed with the boot we're at
+		int tileAt = 0;
+		// the min step a boot needs to be able to take to reach the barn
+		int neededStep = 1;
+		boolean[] canReach = new boolean[bootNum];
+		for (Boot b : boots) {
+			/*
+			 * remove all the tiles that this boot can't traverse
+			 * & update tile_at & needed_step accordingly
+			 */
+			while (tileAt < tilesByDepth.length
+					&& tiles[tilesByDepth[tileAt]] > b.maxDepth) {
+				int invalid = tilesByDepth[tileAt];
+				validTiles.remove(invalid);
+				neededStep = Math.max(
+					neededStep,
+					validTiles.ceiling(invalid) - validTiles.floor(invalid)
+				);
+				tileAt++;
+			}
+			canReach[b.index] = b.maxSteps >= neededStep;
+		}
+		
+		for (boolean b : canReach) {
+			io.println(b ? 1 : 0);
+		}
 
-	for (bool b: ans)
-		cout << b << "\n";
+		io.close();
+	}
+	
+	//CodeSnip{Kattio}
 }
 ```
+
+</JavaSection>
+<PySection>
+
+```py
+from typing import NamedTuple
+
+class Boot(NamedTuple):
+	max_depth: int
+	max_steps: int
+	index: int
+
+class Tile:
+	def __init__(self, before: int, after: int):
+		self.before = before
+		self.after = after
+
+with open('snowboots.in') as read:
+	tile_num, boot_num = [int(i) for i in read.readline().split()]
+	tiles = [int(i) for i in read.readline().split()]
+	assert tile_num == len(tiles)
```

Comment:
necessary?

Reply:
no, but it's nice to have

Reply:
But in our case (in usaco contest) , can this assertion ever fail?

Reply:
won't fail, but if the user decides to stress test or make their own test cases, this will come into effect

Reply:
Tbh idk if that's something we should take into account about